{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# Extracting GBA carriers\n",
    "\n",
    "Doing this as a request by Jeff Kim, and as a UKBB usage practice."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1",
   "metadata": {},
   "source": [
    "## 0) Just make sure that when you start jupyter instance, start up one with 1tb storage and 8 cores\n",
    "## 1) Download plink exome sequencing data for chr1, chr14, chr19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# plink1 files\n",
    "#!dx ls Bulk/Genotype\\ Results/Genotype\\ calls  # you don't want imputed genotypes - you want exome files instead\n",
    "!dx ls Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ PLINK\\ format\\ -\\ final\\ release/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!dx download Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ PLINK\\ format\\ -\\ final\\ release/ukb23158_c1_b0_v1*\n",
    "!dx download Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ PLINK\\ format\\ -\\ final\\ release/ukb23158_c14_b0_v1*\n",
    "!dx download Bulk/Exome\\ sequences/Population\\ level\\ exome\\ OQFE\\ variants,\\ PLINK\\ format\\ -\\ final\\ release/ukb23158_c19_b0_v1*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4",
   "metadata": {},
   "source": [
    "## 2) Use script to extract E326K, T369M, and N370S carriers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "!echo -e '1:155235843:T:C\\n1:155236246:G:A\\n1:155236376:C:T' > GBA1_risk_variants.snplist\n",
    "#!echo -e 'rs76763715\\nrs75548401\\nrs2230288' > GBA1_risk_variants.snplist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Download plink2\n",
    "!wget https://s3.amazonaws.com/plink2-assets/plink2_linux_x86_64_latest.zip\n",
    "!unzip plink2_linux_x86_64_latest.zip\n",
    "!chmod +x plink2\n",
    "\n",
    "!dx upload plink2 --destination ./"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ONLY NEEDED WHEN USING GENOTYPE DATA\n",
    "# Generate dummy .fam file with placeholder patient metadata (mandatory to run plink command to extract variants)\n",
    "import struct\n",
    "\n",
    "# Paths\n",
    "bed_file = \"ukb22418_c1_b0_v2.bed\"\n",
    "fam_file = \"ukb22418_c1_b0_v2.fam\"\n",
    "\n",
    "# Read the number of samples from the .fam file if it exists,\n",
    "# otherwise count from .bed/.bim. We'll count from .bim here:\n",
    "bim_file = \"ukb22418_c1_b0_v2.bim\"\n",
    "\n",
    "# Count number of lines in .bim (number of variants)\n",
    "with open(bim_file) as f:\n",
    "    n_variants = sum(1 for _ in f)\n",
    "\n",
    "# .bed stores samples in columns, but easier: we can assume you know N samples.\n",
    "# If you a .fam and have no idea, you can use plink --bfile ... --write-samples to get the sample count.\n",
    "#!./plink2 --bfile ukb22418_c1_b0_v2 --write-samples\n",
    "\n",
    "# # For demonstration, let's assume N samples\n",
    "N = n_variants  # <-- replace with your number of samples\n",
    "\n",
    "# # Generate dummy .fam\n",
    "with open(fam_file, \"w\") as f:\n",
    "    for i in range(1, N + 1):\n",
    "        # FID IID PAT MAT SEX PHENOTYPE\n",
    "        line = f\"FAM{i} IND{i} 0 0 0 -9\\n\"\n",
    "        f.write(line)\n",
    "\n",
    "print(f\"Dummy .fam file created with {N} samples.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Checking file integrity\n",
    "!ls -lh ukb22418_c1_b0_v2.bed\n",
    "!md5sum ukb22418_c1_b0_v2.bed  # if a checksum is provided\n",
    "\n",
    "# Hash codes are available here: https://biobank.ndph.ox.ac.uk/ukb/refer.cgi?id=998"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!./plink2 --bfile ukb23158_c1_b0_v1 \\\n",
    "         --extract GBA1_risk_variants.snplist \\\n",
    "         --export A \\\n",
    "         --out GBA1_risk_variants_raw \\\n",
    "         --no-categorical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Alternative below, using bgen files:\n",
    "# !./plink2 --bfile ukb22418_c1_b0_v2 --export bgen-1.2 --out ukb22418_c1_b0_v2\n",
    "\n",
    "# !./plink2 \\\n",
    "#   --bgen ukb22418_c1_b0_v2.bgen ref-first \\\n",
    "#   --sample ukb22418_c1_b0_v2.sample \\\n",
    "#   --extract GBA1_risk_variants.snplist \\\n",
    "#   --export A \\\n",
    "#   --out GBA1_risk_variants_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# If both tentatives above failed, use plink1 (1.9) instead\n",
    "# !wget https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20250819.zip\n",
    "# !unzip plink_linux_x86_64_20250819.zip\n",
    "# !chmod +x plink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# IMPORTANT - Don't forget to save your result into permanent storage!\n",
    "!dx upload GBA1_risk_variants_raw.* --destination ./"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Visualize head of .bed and .bim files, and of result\n",
    "#!head ukb23158_c1_b0_v1.bed # lines are too long\n",
    "!head ukb23158_c1_b0_v1.bim\n",
    "!head GBA1_risk_variants_raw.raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Defining a function to extract information (by Jeff Kim)\n",
    "def extract_alternate_carriers_vectorized(raw_file, include_count=False):\n",
    "    \"\"\"\n",
    "    Process PLINK2 raw format (--recode A) to extract participants carrying alternate alleles\n",
    "    using vectorized operations for improved performance.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    raw_file : str\n",
    "        Path to the PLINK2 raw format file\n",
    "    include_count : bool, default=False\n",
    "        If True, output will include a COUNT column with allele counts\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    pd.DataFrame\n",
    "        DataFrame with columns:\n",
    "        - IID: Sample identifier\n",
    "        - VARID: Comma-separated list of variants with alternate alleles\n",
    "        - COUNT: (Only if include_count=True) Comma-separated list of allele counts\n",
    "                 corresponding to each variant in VARID\n",
    "                 0 = Homozygous alternate, 1 = Heterozygous\n",
    "    \"\"\"\n",
    "    # Read the raw file\n",
    "    df = pd.read_csv(raw_file, sep = r'\\s+')\n",
    "    \n",
    "    # Get the variant IDs (column names after the first 6 columns)\n",
    "    variant_cols = df.columns[6:]\n",
    "    \n",
    "    # Pre-clean variant names (remove _REF part) - only do this operation once\n",
    "    clean_variants = np.array([var.split('_')[0] for var in variant_cols])\n",
    "    \n",
    "    # Extract genotype data as numpy array for faster processing\n",
    "    genotype_array = df[variant_cols].values\n",
    "    \n",
    "    # Get IIDs as numpy array\n",
    "    iids = df['IID'].values\n",
    "    \n",
    "    # Initialize result dictionaries\n",
    "    result_dict = {}\n",
    "    count_dict = {}\n",
    "    \n",
    "    # Identify alternate allele carriers (value < 2)\n",
    "    alt_allele_mask = genotype_array < 2\n",
    "    \n",
    "    # Process each sample\n",
    "    for i in range(len(df)):\n",
    "        # Find variants where this sample has alternate alleles\n",
    "        alt_indices = np.where(alt_allele_mask[i])[0]\n",
    "        \n",
    "        # Only include samples with at least one alternate allele\n",
    "        if len(alt_indices) > 0:\n",
    "            # Get the variant IDs\n",
    "            alternate_variants = clean_variants[alt_indices]\n",
    "            \n",
    "            # Join variant IDs\n",
    "            result_dict[iids[i]] = ','.join(alternate_variants)\n",
    "            \n",
    "            if include_count:\n",
    "                # Get the actual count values (0 or 1) for these variants\n",
    "                count_values = genotype_array[i, alt_indices]\n",
    "                # Convert to strings and join\n",
    "                count_dict[iids[i]] = ','.join(map(str, count_values))\n",
    "    \n",
    "    # Convert the dictionaries to a DataFrame\n",
    "    if include_count:\n",
    "        result_df = pd.DataFrame({\n",
    "            'IID': list(result_dict.keys()),\n",
    "            'VARID': list(result_dict.values()),\n",
    "            'COUNT': [count_dict[iid] for iid in result_dict.keys()]\n",
    "        })\n",
    "    else:\n",
    "        result_df = pd.DataFrame(list(result_dict.items()), \n",
    "                               columns=['IID', 'VARID'])\n",
    "    \n",
    "    return result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GBA1_carriers = extract_alternate_carriers_vectorized(\"GBA1_risk_variants_raw.raw\", include_count=True)\n",
    "GBA1_E326K_carriers = GBA1_carriers[GBA1_carriers['VARID'].str.contains('1:155236376:C:T')]\n",
    "GBA1_noE326K_carriers = GBA1_carriers[GBA1_carriers['VARID'] != ('1:155236376:C:T')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GBA1_carriers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GBA1_E326K_carriers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GBA1_noE326K_carriers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19",
   "metadata": {},
   "source": [
    "## 3) Use cohort browser to establish EUR PD cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# First, see what is the information in the .fam files\n",
    "!head ukb23158_c1_b0_v1.fam # most probably just sex information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# In cohort browser:\n",
    "# Option 1: Health related outcomes -> First occurrences -> Nervous system disorders -> Date G20 first reported (parkinson's disease)\n",
    "# Option 2 (better): Health related outcomes -> First occurrences -> Nervous system disorders -> Source of report of G20 (parkinson's disease)\n",
    "\n",
    "# Include everything EXCEPT \"Self-report only\" and missing values (there should be around 4,538 participants with PD)\n",
    "\n",
    "# Then, download subject ids and upload here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22",
   "metadata": {},
   "source": [
    "## 4) Find overlaps between #2 and #3 to identify GBA-PD cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "PD_patients = pd.read_csv(\"./UKBB_participant_IDs_w_PD.csv\")\n",
    "PD_patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GBA1_PD_carriers = GBA1_carriers[GBA1_carriers['IID'].isin(PD_patients['Participant ID'])]\n",
    "GBA1_PD_carriers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GBA1_PD_E326K_carriers = GBA1_E326K_carriers[GBA1_E326K_carriers['IID'].isin(PD_patients['Participant ID'])]\n",
    "GBA1_PD_E326K_carriers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "GBA1_PD_noE326K_carriers = GBA1_noE326K_carriers[GBA1_noE326K_carriers['IID'].isin(PD_patients['Participant ID'])]\n",
    "GBA1_PD_noE326K_carriers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 5) Use plink2 --freq count to get allele counts of 14:74480701:C:T and 19:3548233:A:G (NPC2 and MFSD12 variants)\n",
    "Trying to see if MFSD12 and NPC2 have different signals depending on E326K vs other GBA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!./plink2 --bfile ukb23158_c14_b0_v1 --keep GBA1_risk_variants_raw.raw --snp \"14:74480701:C:T\" --freq --out GBA_NPC2_carriers --no-categorical\n",
    "!echo 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxx'\n",
    "!./plink2 --bfile ukb23158_c19_b0_v1 --keep GBA1_risk_variants_raw.raw --snp \"19:3548233:A:G\" --freq --out GBA_MFSD12_carriers --no-categorical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!head GBA_NPC2_carriers.afreq\n",
    "!head GBA_MFSD12_carriers.afreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Do the same, but for PD subjects only: all GBA, E326K and no E326K\n",
    "\n",
    "# # First, save all PD subjects with stratified GBA\n",
    "# GBA1_PD_carriers.to_csv(\"GBA1_PD_carriers.txt\", sep=\"\\t\", index=False)\n",
    "# !dx upload GBA1_PD_carriers.txt\n",
    "\n",
    "# GBA1_PD_E326K_carriers.to_csv(\"GBA1_PD_E326K_carriers.txt\", sep=\"\\t\", index=False)\n",
    "# !dx upload GBA1_PD_E326K_carriers.txt\n",
    "\n",
    "# GBA1_PD_noE326K_carriers.to_csv(\"GBA1_PD_noE326K_carriers.txt\", sep=\"\\t\", index=False)\n",
    "# !dx upload GBA1_PD_noE326K_carriers.txt\n",
    "\n",
    "# GBA1_PD_carriers\n",
    "keep_df = pd.DataFrame({\n",
    "    \"FID\": GBA1_PD_carriers[\"IID\"],\n",
    "    \"IID\": GBA1_PD_carriers[\"IID\"]\n",
    "})\n",
    "keep_df.to_csv(\"GBA1_PD_subjects.txt\", sep=\"\\t\", index=False, header=False)\n",
    "!dx upload GBA1_PD_subjects.txt\n",
    "\n",
    "# GBA1_PD_E326K_carriers\n",
    "keep_df = pd.DataFrame({\n",
    "    \"FID\": GBA1_PD_E326K_carriers[\"IID\"],\n",
    "    \"IID\": GBA1_PD_E326K_carriers[\"IID\"]\n",
    "})\n",
    "keep_df.to_csv(\"GBA1_PD_E326K_subjects.txt\", sep=\"\\t\", index=False, header=False)\n",
    "!dx upload GBA1_PD_E326K_subjects.txt\n",
    "\n",
    "# GBA1_PD_noE326K_carriers\n",
    "keep_df = pd.DataFrame({\n",
    "    \"FID\": GBA1_PD_noE326K_carriers[\"IID\"],\n",
    "    \"IID\": GBA1_PD_noE326K_carriers[\"IID\"]\n",
    "})\n",
    "keep_df.to_csv(\"GBA1_PD_noE326K_subjects.txt\", sep=\"\\t\", index=False, header=False)\n",
    "!dx upload GBA1_PD_noE326K_subjects.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Check if IDs match\n",
    "!head ukb23158_c14_b0_v1.fam\n",
    "!head GBA1_PD_subjects.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## PD, all GBA\n",
    "!./plink2 --bfile ukb23158_c14_b0_v1 --keep GBA1_PD_subjects.txt --snp \"14:74480701:C:T\" --freq --out GBA_PD_subjects_NPC2 --no-pheno\n",
    "!echo 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxx'\n",
    "!./plink2 --bfile ukb23158_c19_b0_v1 --keep GBA1_PD_subjects.txt --snp \"19:3548233:A:G\" --freq --out GBA_PD_subjects_MFSD12 --no-pheno\n",
    "\n",
    "## PD, E326K only\n",
    "!./plink2 --bfile ukb23158_c14_b0_v1 --keep GBA1_PD_E326K_subjects.txt --snp \"14:74480701:C:T\" --freq --out GBA_PD_E326K_subjects_NPC2 --no-pheno\n",
    "!echo 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxx'\n",
    "!./plink2 --bfile ukb23158_c19_b0_v1 --keep GBA1_PD_E326K_subjects.txt --snp \"19:3548233:A:G\" --freq --out GBA_PD_E326K_subjects_MFSD12 --no-pheno\n",
    "\n",
    "## PD, no E326K\n",
    "!./plink2 --bfile ukb23158_c14_b0_v1 --keep GBA1_PD_noE326K_subjects.txt --snp \"14:74480701:C:T\" --freq --out GBA_PD_noE326K_subjects_NPC2 --no-pheno\n",
    "!echo 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxx'\n",
    "!./plink2 --bfile ukb23158_c19_b0_v1 --keep GBA1_PD_noE326K_subjects.txt --snp \"19:3548233:A:G\" --freq --out GBA_PD_noE326K_subjects_MFSD12 --no-pheno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!head GBA_PD_subjects_NPC2.afreq\n",
    "!head GBA_PD_subjects_MFSD12.afreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!head GBA_PD_E326K_subjects_NPC2.afreq\n",
    "!head GBA_PD_E326K_subjects_MFSD12.afreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!head GBA_PD_noE326K_subjects_NPC2.afreq\n",
    "!head GBA_PD_noE326K_subjects_MFSD12.afreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!dx upload *.afreq ./"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37",
   "metadata": {},
   "source": [
    "## 6) Do the same for non-PD carriers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## No PD-GBA\n",
    "!./plink2 --bfile ukb23158_c14_b0_v1 --remove GBA1_PD_subjects.txt --snp \"14:74480701:C:T\" --freq --out GBA_Ctrl_subjects_NPC2 --no-pheno\n",
    "!echo 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxx'\n",
    "!./plink2 --bfile ukb23158_c19_b0_v1 --remove GBA1_PD_subjects.txt --snp \"19:3548233:A:G\" --freq --out GBA_Ctrl_subjects_MFSD12 --no-pheno\n",
    "\n",
    "## No PD-GBA-E326K\n",
    "!./plink2 --bfile ukb23158_c14_b0_v1 --remove GBA1_PD_E326K_subjects.txt --snp \"14:74480701:C:T\" --freq --out GBA_Ctrl_E326K_subjects_NPC2 --no-pheno\n",
    "!echo 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxx'\n",
    "!./plink2 --bfile ukb23158_c19_b0_v1 --remove GBA1_PD_E326K_subjects.txt --snp \"19:3548233:A:G\" --freq --out GBA_Ctrl_E326K_subjects_MFSD12 --no-pheno\n",
    "\n",
    "## No PD-GBA-noE326K\n",
    "!./plink2 --bfile ukb23158_c14_b0_v1 --remove GBA1_PD_noE326K_subjects.txt --snp \"14:74480701:C:T\" --freq --out GBA_Ctrl_noE326K_subjects_NPC2 --no-pheno\n",
    "!echo 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxx'\n",
    "!./plink2 --bfile ukb23158_c19_b0_v1 --remove GBA1_PD_noE326K_subjects.txt --snp \"19:3548233:A:G\" --freq --out GBA_Ctrl_noE326K_subjects_MFSD12 --no-pheno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!head GBA_Ctrl_subjects_NPC2.afreq\n",
    "!head GBA_Ctrl_subjects_MFSD12.afreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!head GBA_Ctrl_E326K_subjects_NPC2.afreq\n",
    "!head GBA_Ctrl_E326K_subjects_MFSD12.afreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!head GBA_Ctrl_noE326K_subjects_NPC2.afreq\n",
    "!head GBA_Ctrl_noE326K_subjects_MFSD12.afreq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The approach above still included some non-GBA PD cases. Have to go back a few steps.\n",
    "# Leverage the following function, to exclude all subjects with PD from complete list of UKBB subjects:\n",
    "#GBA1_Ctrl_carriers = GBA1_carriers[~GBA1_carriers['IID'].isin(PD_patients['Participant ID'])]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
