{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing data for haplotype analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This consists of the following basic steps:\n",
    "- Subset by ACAF threshold\n",
    "- Subset by desired region (SPTSSB lead SNP 500kb before and after)\n",
    "- Phase data\n",
    "- Filter variants by p < 0.05 or 0.03 (PD GWAS stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Useful packages and variables\n",
    "\n",
    "import os                                       # interact with the environment\n",
    "import sys                                      # interact with stderr\n",
    "from firecloud import api as fapi               # interact with the workspace\n",
    "from io import StringIO                         # work with file contents\n",
    "import pandas as pd                             # work with tabular data\n",
    "import numpy as np\n",
    "import urllib.parse                             # build URLs to Google Cloud Console\n",
    "from IPython.core.display import display, HTML  # display links to Google Cloud Console\n",
    "\n",
    "# Enable IPython to display matplotlib graphs.\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "from cycler import cycler\n",
    "\n",
    "# Get workspace attributes, so the notebook can access workspace data\n",
    "BILLING_PROJECT_ID = os.environ['GOOGLE_PROJECT']\n",
    "WORKSPACE_NAMESPACE = os.environ['WORKSPACE_NAMESPACE']\n",
    "WORKSPACE_NAME = os.environ['WORKSPACE_NAME']\n",
    "WORKSPACE_BUCKET = os.environ['WORKSPACE_BUCKET']\n",
    "\n",
    "WORKSPACE_ATTRIBUTES = fapi.get_workspace(WORKSPACE_NAMESPACE, WORKSPACE_NAME).json().get('workspace',{}).get('attributes',{})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create folder\n",
    "#!mkdir haplotype_analysis\n",
    "%cd haplotype_analysis/\n",
    "!ls -l\n",
    "#!mkdir inputed_genotypes\n",
    "#!mv inputed_genotypes/ imputed_genotypes/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Copy imputed genotype for chromosome of interest\n",
    "#!gsutil -mu {BILLING_PROJECT_ID} ls gs://gp2tier2/release10/imputed_genotypes/EUR/\n",
    "!gsutil -mu {BILLING_PROJECT_ID} cp gs://gp2tier2/release10/imputed_genotypes/EUR/chr3_EUR_release10.pgen imputed_genotypes/\n",
    "!ls -l imputed_genotypes/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Install plink (only need to run once)\n",
    "%cd ../../\n",
    "!wget https://s3.amazonaws.com/plink2-assets/plink2_linux_x86_64_latest.zip\n",
    "!unzip plink2_linux_x86_64_latest.zip\n",
    "!chmod +x plink2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ACAF filtering\n",
    "ACAF = Ancestry-Corrected Allele Frequency, usually calculated within a homogeneous ancestry group \n",
    "(e.g., EUR, EAS, AFR). It helps avoid population stratification artifacts.\n",
    "\n",
    "Mostly important if dealing with very large WGS files from All of Us."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Assuming .afreq are not available - Need to generate first\n",
    "#%cd haplotype_analysis/imputed_genotypes/\n",
    "!ls -l\n",
    "#!head chr3_EUR_release10.psam\n",
    "\n",
    "# Calculate allele frequency within EUR group -> .afreq file\n",
    "!../../plink2 \\\n",
    "  --pfile chr3_EUR_release10 \\\n",
    "  --freq \\\n",
    "  --out chr3_eur_freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Generate list of SNPs\n",
    "#!head chr3_eur_freq.afreq\n",
    "!awk '$5 > 0.01 && $5 < 0.99' chr3_eur_freq.afreq | cut -f2 > acaf_filtered_snps.txt\n",
    "!head acaf_filtered_snps.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter based on ACAF threshold\n",
    "!../../plink2 \\\n",
    " --pfile chr3_EUR_release10 \\\n",
    " --extract acaf_filtered_snps.txt \\\n",
    " --make-pgen \\\n",
    " --out chr3_acaf_filtered\n",
    "\n",
    "#!head chr3_eur_freq.afreq\n",
    "!head chr3_acaf_filtered.pvar\n",
    "#!gsutil -u {BILLING_PROJECT_ID} cat gs://gp2tier2/release10/README_release10_01072025.txt | head -n 20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subsetting by locus of interest (SPTSSB)\n",
    "Genome assembly is hg38 - chr3:160859842-161859842;\n",
    "Coordinates in hg19 - chr3:160577630-161577630"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Subset to region of interest (SPTSSB) - Same coordinates used in coloc analysis\n",
    "!../../plink2 \\\n",
    "  --pfile chr3_acaf_filtered \\\n",
    "  --chr 3 \\\n",
    "  --from-bp 160859842 \\\n",
    "  --to-bp 161859842 \\\n",
    "  --make-pgen \\\n",
    "  --out chr3_acaf_filtered_sptssb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phasing data\n",
    "In theory, you could generate .ped and .map files now with the command below. But it is recommend to do phasing first instead\n",
    "\n",
    "plink2 --pfile chr3_acaf_filtered_sptssb --recode --out sptssb_haplo\n",
    "\n",
    "Phasing is the process of determining which alleles at multiple SNPs are inherited together on the same chromosome copy \n",
    "(i.e., haplotype resolution). In other words, it determines which allele came from which parent.\n",
    "\n",
    "For that, we will use a well-established tool in the field called SHAPEIT, using 1000G or TOPMed reference map."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Convert plink files to vcf\n",
    "!../../plink2 \\\n",
    "  --pfile chr3_acaf_filtered_sptssb \\\n",
    "  --recode vcf bgz \\\n",
    "  --out chr3_acaf_filtered_sptssb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Download genetic map (only need to run once)\n",
    "!wget https://github.com/odelaneau/shapeit4/raw/master/maps/genetic_maps.b38.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Download reference panel (only once), for improved accuracy. 1000G files are in hg19 (need to liftOver later)\n",
    "!wget https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr3.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz  # 1000G\n",
    "!wget https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr3.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz.tbi  # 1000G\n",
    "#!wget ftp://ngs.sanger.ac.uk/production/hg38/1000G/1000G.GRCh38.autosomes.genotypes.20170504/ALL.chr3.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz\n",
    "\n",
    "!ls -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Install SHAPEIT4 or 5 (in that order - only once)\n",
    "# Easier to do in the terminal.\n",
    "\n",
    "# In case wanting to use a more recent version of GATK\n",
    "#!wget https://github.com/broadinstitute/gatk/releases/download/4.5.0.0/gatk-4.5.0.0.zip # gatk\n",
    "#!unzip gatk-4.5.0.0.zip\n",
    "\n",
    "#!git clone https://github.com/odelaneau/shapeit4.git\n",
    "#!git clone https://github.com/odelaneau/shapeit5.git\n",
    "!wget https://github.com/odelaneau/shapeit5/releases/download/v5.1.1/phase_common_static\n",
    "\n",
    "!ls -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!chmod 777 ./phase_common_static\n",
    "!./phase_common_static"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Rename chromosomes in vcfs, so they match the fasta reference genome\n",
    "# Create \"chr_rename.txt\"\n",
    "# with open(\"chr_rename.txt\", \"w\") as f:\n",
    "#     # Autosomes\n",
    "#     for i in range(1, 23):\n",
    "#         f.write(f\"{i}\\tchr{i}\\n\")\n",
    "#     # Sex chromosomes + MT\n",
    "#     f.write(\"X\\tchrX\\n\")\n",
    "#     f.write(\"Y\\tchrY\\n\")\n",
    "#     f.write(\"MT\\tchrM\\n\")\n",
    "\n",
    "# Run renaming itself\n",
    "!bcftools annotate \\\n",
    "  --rename-chrs chr_rename.txt \\\n",
    "  -o ALL.chr3.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes_renamed.vcf.gz \\\n",
    "  -O z \\\n",
    "  ALL.chr3.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz\n",
    "\n",
    "!bcftools index ALL.chr3.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes_renamed.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## LiftOver 1000G reference panel, originally in hg19\n",
    "# Download required files (only once)\n",
    "#!wget https://hgdownload.cse.ucsc.edu/goldenpath/hg19/liftOver/hg19ToHg38.over.chain.gz\n",
    "#!wget https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa # used by GATK\n",
    "#!wget https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa.fai # used by GATK\n",
    "#!wget https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.dict # used by GATK\n",
    "\n",
    "#!ls -l\n",
    "\n",
    "# Run the liftOver itself\n",
    "# CONTINUE TROUBLESHOOTING FROM HERE\n",
    "# Since providing a reference panel is not mandatory for SHAPEIT, we will skip it for now.\n",
    "\n",
    "!gatk --java-options \"-Xmx8g\" LiftoverVcf \\\n",
    "  -I ALL.chr3.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes_renamed.vcf.gz \\\n",
    "  -O chr3_lifted_to_hg38.vcf.gz \\\n",
    "  -CHAIN hg19ToHg38.over.chain.gz \\\n",
    "  -REJECT chr3_rejected_variants.vcf.gz \\\n",
    "  -R GRCh38_full_analysis_set_plus_decoy_hla.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Forgot to create index\n",
    "!tabix haplotype_analysis/imputed_genotypes/chr3_acaf_filtered_sptssb.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create AC field (required by SHAPEIT)\n",
    "!bcftools +fill-tags haplotype_analysis/imputed_genotypes/chr3_acaf_filtered_sptssb.vcf.gz -o haplotype_analysis/imputed_genotypes/chr3_acaf_filtered_sptssb_filled.vcf.gz\n",
    "\n",
    "!tabix haplotype_analysis/imputed_genotypes/chr3_acaf_filtered_sptssb_filled.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unzip genetic map\n",
    "!tar -xzvf genetic_maps.b38.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run SHAPEIT5 itself\n",
    "!./phase_common_static \\\n",
    "  --input haplotype_analysis/imputed_genotypes/chr3_acaf_filtered_sptssb_filled.vcf.gz \\\n",
    "  --map chr3.b38.gmap.gz \\\n",
    "  --region 3:160859842-161859842 \\\n",
    "  --output sptssb_phased.bcf \\\n",
    "  --log sptssb_shapeit5.log\n",
    "\n",
    "# OPTIONAL\n",
    "#  --reference chr3_lifted_to_hg38.vcf.gz \\\n",
    "\n",
    "!ls -l\n",
    "\n",
    "# Verify the output\n",
    "#!bcftools view sptssb_phased.vcf.gz | head -n 20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering GWAS variants\n",
    "Will initially do a \"soft\" filtering using all PD GWAS variants with p<0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Preparing PD GWAS file (case/control)\n",
    "# File below was uploaded manually. Can be found in the following link: https://api.kpndataregistry.org/api/d/7j5797\n",
    "\n",
    "!gsutil -mu {BILLING_PROJECT_ID} cat gs://fc-0e367d5c-f21f-4d17-9f20-16910ff5c3aa/uploads/GP2_GWAS_2025/GP2_ALL_EUR_CLINICAL_ONLY_HG38_12162024.txt.gz | zcat | cut -f 1,2,8 | grep -P '^3' > GP2_GWAS_case_ctrl_chr3.txt\n",
    "!sed -i '1iCHROM\\tPOS\\tGWAS_P' GP2_GWAS_case_ctrl_chr3.txt\n",
    "\n",
    "# Replace multiple spaces with tabs\n",
    "!awk '{$1=$1}1' OFS='\\t' GP2_GWAS_case_ctrl_chr3.txt > GP2_GWAS_case_ctrl_chr3.tsv\n",
    "\n",
    "!head GP2_GWAS_case_ctrl_chr3.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compreess and index annotation\n",
    "!bgzip -c GP2_GWAS_case_ctrl_chr3.tsv > GP2_GWAS_case_ctrl_chr3.tsv.gz\n",
    "!tabix -s 1 -b 2 -e 2 -H GP2_GWAS_case_ctrl_chr3.tsv.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Option 1 (failed) - Annotate BCF files with GWAS p-values, then filter\n",
    "# # Annotate\n",
    "# !bcftools annotate \\\n",
    "#   -a GP2_GWAS_case_ctrl_chr3.tsv.gz \\\n",
    "#   -c CHROM,POS,GWAS_P \\\n",
    "#   -h header.txt \\\n",
    "#   -O b -o sptssb_phased_annotated.bcf \\\n",
    "#   sptssb_phased.bcf\n",
    "\n",
    "# # Filter annotated VCF for p<0.05\n",
    "# !bcftools view \\\n",
    "#   -i 'INFO/GWAS_P<0.05' \\\n",
    "#   -O b -o sptssb_phased_filtered.bcf \\\n",
    "#   sptssb_phased_annotated.bcf\n",
    "\n",
    "# # Check if output is correct\n",
    "# !bcftools view -H sptssb_phased_filtered.bcf | more"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Option 2 (worked) - Filter using regions file\n",
    "# Create regions file - only significant PD GWAS variants\n",
    "#!awk 'NR>1 && $3 < 0.05 { print $1\"\\t\"$2 }' GP2_GWAS_case_ctrl_chr3.tsv > significant_variants.tsv\n",
    "\n",
    "#!gsutil -mu {BILLING_PROJECT_ID} cp gs://fc-0e367d5c-f21f-4d17-9f20-16910ff5c3aa/uploads/GP2_GWAS_2025/GP2_GWAS_SPTSSB_vars_5x10-5_+-5kb.txt ./\n",
    "#!gsutil -mu {BILLING_PROJECT_ID} cp gs://fc-0e367d5c-f21f-4d17-9f20-16910ff5c3aa/uploads/GP2_GWAS_2025/GP2_GWAS_SPTSSB_vars_5x10-8_+-5kb.txt ./\n",
    "!gsutil -mu {BILLING_PROJECT_ID} cp gs://fc-0e367d5c-f21f-4d17-9f20-16910ff5c3aa/uploads/GP2_GWAS_2025/GP2_GWAS_SPTSSB_vars_Jeff_prioritized.txt ./\n",
    "\n",
    "# Run filtering itself\n",
    "#!bcftools view -T significant_variants.tsv sptssb_phased.bcf -O b -o sptssb_phased_filtered.bcf\n",
    "!bcftools view -T GP2_GWAS_SPTSSB_vars_Jeff_prioritized.txt sptssb_phased.bcf -O b -o sptssb_phased_filtered_4.bcf\n",
    "\n",
    "# Check if output is correct - non-filtered: 3138 variants; filtered: 1447 variants; filtered 2/3/4: 33/24/8\n",
    "!bcftools view -H sptssb_phased.bcf | wc -l\n",
    "!bcftools view -H sptssb_phased_filtered_2.bcf | wc -l\n",
    "!bcftools view -H sptssb_phased_filtered_3.bcf | wc -l\n",
    "!bcftools view -H sptssb_phased_filtered_4.bcf | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Converting files to .ped and .map\n",
    "Those are the file formats required by Mary Makarious haplotype analysis pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Convert .bcf to .vcf\n",
    "!bcftools view -O v -o sptssb_phased_filtered_4.vcf sptssb_phased_filtered_4.bcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run plink to generate files\n",
    "!./plink2 --vcf sptssb_phased_filtered_4.vcf --export ped --out sptssb_phased_filtered_4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (Final Step) Rescue phenotype information\n",
    "In the process of data prep/phasing, phenotype information (sex and case/control status) is often lost.\n",
    "\n",
    "This information is critical for our haplotype analysis, since we want to compare haplotype frequencies between PD cases and healthy controls."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## List sample IDs\n",
    "#!gsutil -mu {BILLING_PROJECT_ID} ls -l gs://gp2tier2/release10/imputed_genotypes/EUR/\n",
    "\n",
    "#!cut -d' ' -f2 yourfile.fam > samples.txt  # IID\n",
    "!head haplotype_analysis/imputed_genotypes/chr3_EUR_release10.psam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Prepare phenotype file (.tsv)\n",
    "# IID     PD_status\n",
    "# ID001   1\n",
    "# ID002   2\n",
    "# 1 is likely control, and 2 is likely case\n",
    "\n",
    "!cut -f 1,2,3 haplotype_analysis/imputed_genotypes/chr3_EUR_release10.psam | sed 's/ /\\t/g' > chr3_pheno_map.tsv\n",
    "!head chr3_pheno_map.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Update .ped\n",
    "#!head sptssb_phased_filtered_4.ped # code -9 means missing phenotype\n",
    "!awk 'NR==FNR {a[$1]=$2; next} { $5=a[$2]; print }' chr3_pheno_map.tsv sptssb_phased_filtered_4.ped > test0.ped\n",
    "!awk 'NR==FNR {a[$1]=$3; next} { $6=a[$2]; print }' chr3_pheno_map.tsv test0.ped > test.ped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Verify new .ped\n",
    "!head test.ped\n",
    "!echo xxxxxx\n",
    "!cut -d' ' -f5 test.ped | sort | uniq -c # sex\n",
    "!echo xxxxxx\n",
    "!cut -d' ' -f6 test.ped | sort | uniq -c # case/control\n",
    "!echo xxxxxx\n",
    "!cut -f2 haplotype_analysis/imputed_genotypes/chr3_EUR_release10.psam | sort | uniq -c # sex\n",
    "!echo xxxxxx\n",
    "!cut -f3 haplotype_analysis/imputed_genotypes/chr3_EUR_release10.psam | sort | uniq -c # case/control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Rename .peds\n",
    "!mv sptssb_phased_filtered_4.ped sptssb_phased_filtered_4_noPheno.ped\n",
    "!mv test.ped sptssb_phased_filtered_4.ped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test - Haplotype analysis with PLINK\n",
    "This approach is more computationally efficient for working with ~30 variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Installing PLINK 1.9\n",
    "#!wget https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20250731.zip\n",
    "!unzip plink_linux_x86_64_20250731.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Running PLINK haplotype analysis - Only works with PLINK 1!\n",
    "!./plink --bfile sptssb_phased_filtered_3.bcf --hap --hap-freq --out sptssb_haplo_freqs_plink"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
