{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# SPTSSB Haplotype Analyses\n",
    "* Pipeline adapted from Mary Makarious: https://github.com/GP2code/India-metaGWAS/blob/main/analyses/09_mapt_haplotype_analysis.ipynb\n",
    "\n",
    "Please run this Jupyter notebook using R as a Kernel.\n",
    "\n",
    "Variants in filtered phased file: 1,447 variants within SPTSSB locus (+-500kb from GP2 GWAS lead SNP), with MAF > 0.01 and GP2 PD GWAS p < 0.05\n",
    "- This number is too high for this pipeline. Instead, from this list use only the variants with PD GWAS p < 5x10^-8 and +-5kb of SPTSSB. -> 24 variants\n",
    "- UPDATE 1: This pipeline is not working for 24 variants. See alternative pipeline named \"_test2\".\n",
    "- UPDATE 2: From visually inspecting the LDhap map, there are some potential \"minimum\" haplotypes (3 variants) that could be tested. Also, I can test 8 variants prioritized by Jeff or ~10 variants in highest LD with lead SNP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#install.packages(\"polspline\")\n",
    "#install.packages(\"https://cran.r-project.org/src/contrib/Archive/rms/rms_6.7-0.tar.gz\", repos = NULL, type = \"source\")\n",
    "install.packages(\"haplo.stats\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Libraries\n",
    "library(haplo.stats)\n",
    "library(ggplot2)\n",
    "library(tidyverse)\n",
    "seed <- c(17, 53, 1, 40, 37, 0, 62, 56, 5, 52, 12, 1)\n",
    "set.seed(seed)\n",
    "seed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "packageVersion('haplo.stats')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "R.version.string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Dirs\n",
    "data_dir='/home/jupyter/SPTSSB_haplotype_analysis/edit/'\n",
    "\n",
    "## Filenames\n",
    "ped_file='sptssb_phased_filtered_4.ped'\n",
    "map_file='sptssb_phased_filtered_4.map'\n",
    "\n",
    "# .ped - Genotype file, containing:\n",
    "# Family ID, Individual ID, Paternal ID, Maternal ID, Sex, Phenotype (first 6 columns)\n",
    "# 1) Extract individuals and SNPs from your master .bed/.bim/.fam PLINK files.\n",
    "# 2) Keep only SNPs located within the SPTSSB gene region, or your region of interest around it.\n",
    "\n",
    "# .map - SNP metadata file, containing:\n",
    "# Chromosome, SNP ID, Genetic distance (set to 0 if unknown), Base-pair position\n",
    "# 1) Match the SNPs in the .ped file, extracted from your main GWAS dataset (likely via PLINK).\n",
    "# 2) Make sure they correspond to SNPs located in the SPTSSB region."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6",
   "metadata": {},
   "source": [
    "## Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### DATA SPTSSB .ped & .map files\n",
    "## MAP file\n",
    "map_data <- read.table(paste0(data_dir, map_file), header = FALSE)\n",
    "## PED File (Genotype Data)\n",
    "ped_data <- read.table(paste0(data_dir, ped_file), header = FALSE)\n",
    "## PCA file\n",
    "#\n",
    "\n",
    "# Extract the genotype data (from the 7th column onwards)\n",
    "geno_data <- ped_data[, 7:ncol(ped_data)]\n",
    "\n",
    "# The Phenotype (case/control phenotype, binary: 0 = control, 1 = case) is in the 6th column\n",
    "phenotype <- ped_data[, 6]\n",
    "# Recode phenotype: 1 -> 0 (Control), 2 -> 1 (Case)\n",
    "phenotype <- ifelse(phenotype == 1, 0, 1)\n",
    "# The sex (covariate) is in the 5th column\n",
    "sex <- ped_data[, 5]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "head(map_data)\n",
    "head(ped_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "head(geno_data, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cat('outcome',phenotype[1:20], '\\n')\n",
    "cat('covariate', sex[1:20] )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11",
   "metadata": {},
   "source": [
    "# Exploratory analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12",
   "metadata": {},
   "source": [
    "## Create a Genotype Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Assign labels to geno data\n",
    "# Create new column names by appending \".a1\" and \".a2\" for each SNP\n",
    "\n",
    "## Add labels to geno matrix\n",
    "snp_names= map_data$V2\n",
    "cat('labels :', snp_names)\n",
    "\n",
    "new_column_names <- c()\n",
    "for (snp in snp_names) {\n",
    "  new_column_names <- c(new_column_names, paste0(snp, \".a1\"), paste0(snp, \".a2\"))\n",
    "}\n",
    "# Assign the new column names to the data frame\n",
    "colnames(geno_data) <- new_column_names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "head(geno_data,4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15",
   "metadata": {},
   "source": [
    "## 2. Estimate Haplotype Frequency with `haplo.em`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## THIS STEP DOES NOT RUN WITH ~20 VARIANTS!\n",
    "save.em <- haplo.em(geno=geno_data, locus.label=snp_names, miss.val=c(0,NA))\n",
    "names(save.em)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "print(save.em)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dim(save.em)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19",
   "metadata": {},
   "source": [
    "## Remark\n",
    "\n",
    "* The print methods shows the haplotypes and their estimated frequencies, followed by the final log-likelihood\n",
    "statistic and the lr stat for no LD, which is the likelihood ratio test statistic contrasting the lnlike for the\n",
    "estimated haplotype frequencies versus the lnlike under the null assuming that alleles from all loci are in\n",
    "linkage equilibrium."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20",
   "metadata": {},
   "source": [
    "### Summary method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "summary(save.em)#, nlines=7)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22",
   "metadata": {},
   "source": [
    "### Remark\n",
    "* The first part of the summary output lists the subject id (row number of input geno matrix), the codes for the haplotypes of each pair, and the posterior probabilities of the haplotype pairs.\n",
    "* The second part gives a table of the maximum number of pairs of haplotypes per subject, versus the number of pairs used\n",
    "in the final posterior probabilities. \n",
    "*  The haplotype codes remove the clutter of illustrating all the alleles of the haplotypes, but may not be as informative as the actual haplotypes themselves.\n",
    "* To see the actual haplotypes, use the show.haplo=TRUE option, as in the following example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# show full haplotypes, instead of codes\n",
    "summary(save.em, show.haplo=TRUE, nlines=7)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24",
   "metadata": {},
   "source": [
    "## Haplotype Frequencies by Group Subsets using `haplo.group`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#help(haplo.group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Frequencies\n",
    "# Group 0 is control, group 1 is PD\n",
    "group.bin=haplo.group(phenotype, geno_data, locus.label=snp_names, miss.val=0)\n",
    "print(group.bin, nlines=15)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27",
   "metadata": {},
   "source": [
    "### Remark\n",
    "* The group.bin object can be very large, depending on the number of possible haplotypes, so only a portion of the output is illustrated above (limited again by nlines). The first section gives a short summary of how many subjects appear in each of the groups. The second section is a table with the following columns:\n",
    "    *  The first column gives row numbers.\n",
    "    *  Total are the estimated haplotype frequencies for the entire data set.\n",
    "    *  The last columns are the estimated haplotype frequencies for the subjects in the levels of the group variable (phenotye.0 and phenotype.1).\n",
    "    *  Note that some haplotype frequencies have an NA, which appears when the haplotypes do not occur in the subgroups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "class(group.bin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "head(group.bin$group.df,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cat('Number of controls and cases', group.bin$group.count,'\\n')\n",
    "cat('Number of SPTSSB loci: ', group.bin$n.loci, '\\n' )\n",
    "freq_haplotypes= group.bin$group.df\n",
    "cat('Total number of SPTSSB haplotypes', dim(freq_haplotypes), '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "group.bin$group.count"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32",
   "metadata": {},
   "source": [
    "**Note:**\n",
    "* A total of 16 haplotypes created from the 8 SPTSSB variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "summary(freq_haplotypes$Total)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "freq_haplotypes[freq_haplotypes$Total > 0.01,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dim(freq_haplotypes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36",
   "metadata": {},
   "source": [
    "## Regression Models: `haplo.glm`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37",
   "metadata": {},
   "source": [
    "### Preparing the data.frame for haplo.glm\n",
    "* A data.frame must be defined, and this object must contain the trait and other optional covariates, plus a special kind of genotype matrix (`geno.glm\n",
    "\n",
    "* Below we prepare a genotype matrix, `geno.glm`, and create a data.frame object, `glm.data`, for use in haplo.glm."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Set up data for haplo.glm, include geno.glm,\n",
    "# covariates sex and respnose is phenotype (PD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Set up data for haplo.glm, include geno.glm,\n",
    "# covariates age and male, and responses resp and y.bin\n",
    "geno.glm <- setupGeno(geno_data,  miss.val=c(0,NA), locus.label=snp_names)\n",
    "attributes(geno.glm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Combine Genotype data, covariates and respose for model fit Model fit\n",
    "glm.data <- data.frame(geno.glm, sex=sex, pd_case=phenotype) # combine geno + pheno + covariates for _glm_ analysis\n",
    "length(phenotype)\n",
    "attributes(glm.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "recoded_geno <- setupGeno(geno_data, locus.label = snp_names)\n",
    "head(recoded_geno, 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "head(geno_data, 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43",
   "metadata": {},
   "source": [
    "# Haplotype association analysis with PD (logistic regression)\n",
    "* Model is: fit.pd <- glm(casepd_ ~ sex + hapN, data = data, family =”binomial”)\n",
    "* Covariates: `sex`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "print(dim(geno.glm))\n",
    "head(geno.glm, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "print(dim(geno.glm))\n",
    "head(glm.data, 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46",
   "metadata": {},
   "source": [
    "### i) Haplotype analysis: _PD ~ sex + haplotype + "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Haplotype glm fit plux sex as covariate,\n",
    "# return model matrix\n",
    "#fm <- glm(case ~ age + sex + hap1, data = data, family =”binomial”)\n",
    "fit.pd <- haplo.glm(pd_case ~ sex + geno.glm, family = binomial, data=glm.data, na.action = \"na.geno.keep\",\n",
    "                     locus.label=snp_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "summary(fit.pd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#help(haplo.glm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cat('base haplotype index: ', fit.pd$haplo.base, '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cat('Number of common haplotypes',length(fit.pd$haplo.common), '\\n')\n",
    "cat('Common haplotypes', fit.pd$haplo.common)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "length(fit.pd$coefficients)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53",
   "metadata": {},
   "source": [
    "### Get Haplotype Frequencies for Cases and Controls and Compute Confidence Intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "summary_fit=(summary(fit.pd))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "print(dim(fit.pd$haplo.unique))\n",
    "head(fit.pd$haplo.unique, 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Merge UNIQUE Haplotypes with Haplotype Frequencies\n",
    "haplo_unique <- fit.pd$haplo.unique   # Matrix of unique haplotypes\n",
    "haplo_freq <- fit.pd$haplo.freq       # Numeric vector of haplotype frequencies\n",
    "\n",
    "# Convert haplo_unique to a data frame and assign SNP names as column names\n",
    "haplo_unique_df <- as.data.frame(haplo_unique, stringsAsFactors = FALSE)\n",
    "colnames(haplo_unique_df) <- colnames(haplo_unique)\n",
    "\n",
    "# Add haplotype frequencies as a new column\n",
    "haplo_unique_df$frequency <- haplo_freq\n",
    "\n",
    "# View Final data frame\n",
    "#head(haplo_unique_df,3)\n",
    "print(dim(haplo_unique_df))\n",
    "\n",
    "## Add Haplotype frequencies for cases and controls\n",
    "merged_haplo_data= merge(haplo_unique_df, freq_haplotypes, by= colnames(haplo_unique), all.x = TRUE)\n",
    "cat('Total number of unique haplotypes: ', dim(merged_haplo_data)[1], '\\n')\n",
    "\n",
    "merged_haplo_data <- merged_haplo_data %>%\n",
    "  rename(freq_controls = `phenotype=0`, freq_cases = `phenotype=1`)\n",
    "merged_haplo_data <- merged_haplo_data %>%\n",
    "    mutate(haplo_index=rownames(merged_haplo_data))\n",
    "\n",
    "### Compute Confidence Intervals\n",
    "## First format summary stats\n",
    "sum_stats <- summary_fit$coefficients\n",
    "sum_stats <- data.frame(\n",
    "  Variable = rownames(sum_stats), # Add row names as a new column called \"Variable\"\n",
    "  sum_stats,                      # Add the rest of the data\n",
    "  row.names = NULL                # Remove the row names\n",
    ")\n",
    "\n",
    "### Calculate the Odds Ratio and Confidence Intervals\n",
    "# -Thecoef values represent log-odds (logarithm of the odds ratio) rather than the odds ratios themselves.\n",
    "# -To interpret them as odds ratios, you need to exponentiate the coefficients.\n",
    "#- Converting Coefficients to Odds Ratios\n",
    "sum_stats <- sum_stats %>%\n",
    "  mutate(\n",
    "    OR = exp(coef),                            # Odds Ratio\n",
    "    CI_lower = exp(coef - 1.96 * se),           # Lower 95% CI\n",
    "    CI_upper = exp(coef + 1.96 * se),            # Upper 95% CI\n",
    "  )\n",
    "\n",
    "### Merge the haplotype frequency data with summary stats\n",
    "## First split the variable column\n",
    "# Transforming the data\n",
    "sum_stats <- sum_stats %>%\n",
    "  mutate(\n",
    "    haplo_index = ifelse(\n",
    "      Variable == \"(Intercept)\", \"1001\",\n",
    "      ifelse(Variable == \"sex\", \"1002\",\n",
    "      ifelse(Variable == \"geno.glm.rare\", \"1003\",\n",
    "             as.character(sub(\".*\\\\.(\\\\d+)$\", \"\\\\1\", Variable))\n",
    "      )\n",
    "    )\n",
    "  )\n",
    ")\n",
    "## merge Haplotype  Frequencies and summary stats\n",
    "merged_haplo_stats= merge(merged_haplo_data, sum_stats, by='haplo_index', ) \n",
    "merged_haplo_stats$OR_95_CI <- paste0(round(merged_haplo_stats$OR,2), \" (\", round(merged_haplo_stats$CI_lower, 2), \"-\", round(merged_haplo_stats$CI_upper, 2), \")\")\n",
    "\n",
    "## Apply bonferroni correction to pvalues\n",
    "ntests= nrow(merged_haplo_stats) # Number of tests (Only contains the cmmon haplotypes)\n",
    "# Apply Bonferroni correction\n",
    "merged_haplo_stats$pval_corrected <- pmin(merged_haplo_stats$pval * ntests, 1)\n",
    "Bonferoni_pval=0.05/ntests\n",
    "cat('Number of haplotypes with frequency >= 1%: ', ntests, '\\n')\n",
    "cat('Number to association test performed: ', ntests, '\\n')\n",
    "cat('Bonferroni corrected pvalue Threshold: ', round(Bonferoni_pval, 4),  '\\n')\n",
    "cat('Number of haplotypes significant at p < 0.05: ' , dim(merged_haplo_stats[merged_haplo_stats$pval < 0.05])[1],  '\\n')\n",
    "cat('Number of haplotypes significant at p < ', Bonferoni_pval, ': ', dim(merged_haplo_stats[merged_haplo_stats$pval < Bonferoni_pval])[1],  '\\n')\n",
    "\n",
    "##BASE HAPLOTYPE\n",
    "# ===== haplo.base            A        G         A         C         A      A  0.17716 ==========="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cat('Number of haplotypes significant at p < 0.05: ' , dim(merged_haplo_stats[merged_haplo_stats$pval < 0.05])[1],  '\\n')\n",
    "merged_haplo_stats[merged_haplo_stats$pval < 0.05,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cat('Number of haplotypes significant at p < ', Bonferoni_pval, ': ', dim(merged_haplo_stats[merged_haplo_stats$pval < Bonferoni_pval])[1],  '\\n')\n",
    "merged_haplo_stats[merged_haplo_stats$pval < Bonferoni_pval, ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Create 'tag_h11' and 'tag_h4' columns based on 'chr3:161374774:C:T' values (upstream variant prioritized in African and Ryan Corces analysis)\n",
    "# Step 1: Create tag_h11, tag_h4, and tag_h11c_temp\n",
    "merged_haplo_stats <- merged_haplo_stats %>%\n",
    "  mutate(\n",
    "    tag_h11 = ifelse(`chr3:161374774:C:T` == \"C\", \"H11\", NA),\n",
    "    tag_h4 = ifelse(`chr3:161374774:C:T` == \"T\", \"H4\", NA),\n",
    "    tag_h11c_temp = ifelse(`chr3:161344843:T:C` == \"C\" & `chr3:161359842:A:G` == \"A\", \"H11c\", NA)\n",
    "  )\n",
    "\n",
    "# Step 2: Assign H11c to the row with the maximum freq_cases among those where tag_h11c_temp is H11c\n",
    "# merged_haplo_stats <- merged_haplo_stats %>%\n",
    "#   mutate(\n",
    "#     tag_h11c = ifelse(tag_h1c_temp == \"H11c\" & \n",
    "#                      freq_cases == max(freq_cases[tag_h11c_temp == \"H11c\"], na.rm = TRUE), \"H11c\", NA)\n",
    "#   ) %>%\n",
    "#   select(-tag_h11c_temp) # Drop the temporary column\n",
    "\n",
    "# # Define the labels for tag_h1sub\n",
    "# labels_h11sub <- c('H11b', 'H11d', 'H11e', 'H11f', 'H11g', 'H11i', 'H11j', 'H11k', 'H11m', 'H11n', 'H11o', 'H11p', 'H11q', 'H11r', 'H11x', 'H11y', 'H11z')\n",
    "\n",
    "# # Create the 'tag_h1sub' column based on specified conditions\n",
    "# merged_haplo_stats <- merged_haplo_stats %>%\n",
    "#   # Filter for relevant rows and sort them\n",
    "#   arrange(desc(freq_cases)) %>%\n",
    "#   mutate(tag_h11sub = case_when(\n",
    "#     !is.na(tag_h11) & is.na(tag_h4) & is.na(tag_h11c) ~ labels_h11sub[row_number()],\n",
    "#     TRUE ~ NA_character_\n",
    "#   ))\n",
    "\n",
    "# # Create the 'haplotype_tag' column based on the first non-NA value from tag_h4, tag_h11c, and tag_h11sub\n",
    "# merged_haplo_stats <- merged_haplo_stats %>%\n",
    "#   mutate(haplotype_tag = coalesce(tag_h4, tag_h11c, tag_h11sub))\n",
    "\n",
    "# # Sort the data frame by haplotype_tag\n",
    "# merged_haplo_stats <- merged_haplo_stats %>%\n",
    "#   arrange(haplotype_tag)\n",
    "\n",
    "### Save SPTSSB Happlotype analysis\n",
    "# Save the data frame as a tab-delimited file\n",
    "write.table(merged_haplo_stats, file = \"table_SPTSSB_haplotype_association_with_pd_case_control_Jeff_vars.tsv\", sep = \"\\t\", row.names = FALSE, quote = TRUE)\n",
    "## Save selected columns for paper"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60",
   "metadata": {},
   "source": [
    "#### Save Haplotype Analysis Table\n",
    "* Table: Association between MAPT haplotype and PD risk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Select the specified columns\n",
    "output_sptssb_analysis <- merged_haplo_stats %>%\n",
    "  select(\n",
    "#    haplotype_tag,\n",
    "    `chr3:161344843:T:C`,\n",
    "    `chr3:161359842:A:G`,\n",
    "    `chr3:161366258:A:C`,\n",
    "    `chr3:161372828:C:T`,\n",
    "    `chr3:161374617:T:C`,\n",
    "    `chr3:161374774:C:T`,\n",
    "    `chr3:161375544:G:A`,\n",
    "    `chr3:161376002:CA:C`,\n",
    "    freq_cases,\n",
    "    freq_controls,\n",
    "    OR_95_CI,\n",
    "    pval,\n",
    "    haplo_index,\n",
    "    pval_corrected,\n",
    "    frequency,\n",
    "    se,\n",
    "    coef,\n",
    "    CI_lower,\n",
    "    CI_upper\n",
    "  )\n",
    "\n",
    "# Save the output as a tab-delimited file\n",
    "write.table(output_sptssb_analysis, file = \"table_SPTSSB_haplotype_association_with_pd_case_control_paper_Jeff_vars.tsv\", sep = \"\\t\", row.names = FALSE, quote = TRUE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
